curp/src/server/raw_curp/mod.rs

```diff
@@ -797,7 +808,12 @@ impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
 // Don't grab lock in the following functions(except cb or sp's lock)
 impl<C: 'static + Command, RC: RoleChange + 'static> RawCurp<C, RC> {
     /// Server becomes a candidate
-    fn become_candidate(&self, st: &mut State, cst: &mut CandidateState<C>, log: &Log<C>) -> Vote {
+    fn become_candidate(
+        &self,
+        st: &mut State,
+        cst: &mut CandidateState<C>,
+        mut log: RwLockUpgradableReadGuard<'_, Log<C>>,
```

Comment:
Why should a lock guard be held throughout the entire function, instead of passing the lock directly and lock it when needed?

Reply:
because the lock may use in different level of method, pass the lock directly may require multiple locks, it's difficult to guarantee the locking order, more prone to deadlocks